### PR TITLE
Unify regression loss

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR unified the regression loss in `bbox_head`. 
Attribute `reg_delta_target = True` indicates the regression target is the delta between prediction and target, otherwise regression target is the ground truth bounding box.
Fixed https://github.com/open-mmlab/mmdetection/issues/2143 so that `IoULoss` could be applied directly in `bbox_head`.

### CODE REVIEW

**Comment by yhcao6** [comment]
Currently only supports two-stage model, may also need to support single-stage models.

File: mmdet/models/bbox_heads/bbox_head.py [hunk 1]
```diff
@@ -95,6 +97,25 @@ def get_target(self, sampling_results, gt_bboxes, gt_labels,
             target_stds=self.target_stds)
         return cls_reg_targets
 
+    def get_reg(self, rois, bbox_pred, bbox_targets, img_metas):
+        labels, label_weights, bbox_targets, bbox_weights = bbox_targets
+        if not self.reg_delta_target:
```
Comment (hellock) [inline mmdet/models/bbox_heads/bbox_head.py]: `roi2bbox` may help.

**Comment by hellock** [comment]
#1848

**Comment by xvjiarui** [comment]
Support vanilla IoU type loss in `AnchorHead` is not that elegant.....

**Comment by melikovk** [comment]
Hi @xvjiarui 

It looks great. If I understand correctly, the idea is to set reg_decoded_bbox flag for IOU losses and then to use inverse transform of predictions to calculate loss. One potential drawback of this approach vs my suggestion in #2417 is that you will be back-propagating the inverse transform.

**Comment by xvjiarui** [comment]
> Hi @xvjiarui
> 
> It looks great. If I understand correctly, the idea is to set reg_decoded_bbox flag for IOU losses and then to use inverse transform of predictions to calculate loss. One potential drawback of this approach vs my suggestion in #2417 is that you will be back-propagating the inverse transform.

Yes. It would be BP the inverse transform. For faster R-CNN style method, the outputs of bbox head are always regarded as offset w.r.t to rois. If you would like to regress bbox coordinate directly, a new head should be implemented.

File: mmdet/models/anchor_heads/anchor_head.py [hunk 2]
```diff
@@ -48,6 +48,7 @@ def __init__(self,
                      type='DeltaXYWHBBoxCoder',
                      target_means=(.0, .0, .0, .0),
                      target_stds=(1.0, 1.0, 1.0, 1.0)),
+                 reg_decoded_bbox=False,
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:51]: The docstring is outdated.

**Comment by melikovk** [comment]
@xvjiarui 
I am probably not clearly expressing myself. I am not suggesting to regress to bbox coordinates directly. I am suggesting that IOU is scale invariant and can be used directly with R-CNN outputs w.r.t. to rois with one caveat, that you should not do log transform of w and h, i.e. in bbox2delta instead of doing
```python 
dw = torch.log(gw / pw)
dh = torch.log(gh / ph)
```
you should have an option to do
```python
dw = gw / pw
dh = gh / ph
```
The log transform is unnecessary for IOU type losses (it is meant to control relative contribution from large objects when using L2 or L1 losses). If you have this as an option you will not need to BP inverse transform. It is possible though that the time and memory impact is minor. Thank you for the great work you are doing.

**Comment by xvjiarui** [comment]
Hi @melikovk 
My bad. It is a good idea, which could be easily supported by adding a new `BBoxCoder`. 
This PR is just merged. You are very welcomed to try it out.

**Comment by melikovk** [comment]
Hi @xvjiarui 
This is great. I will test it on my dataset as I am curious to see if IOU or GIOU loss will improve performance. I will give a try to my approach and will let you know if it has any impact. Again, thank you for your great work.
